src/supervisor/watchers/handlers/pagination.ts

```diff
@@ -0,0 +1,92 @@
+import { V1ObjectMeta } from '@kubernetes/client-node';
+import { IncomingMessage } from 'http';
+import * as sleep from 'sleep-promise';
+
+import { calculateSleepSeconds } from '../../kuberenetes-api-wrappers';
+import { V1NamespacedList } from './types';
+import type { IRequestError } from '../../types';
+
+const PAGE_SIZE = 100;
+
+export async function paginatedList<
+  T extends { metadata?: V1ObjectMeta },
+  TList extends { items: Array<T> },
+>(
+  namespace: string,
+  list: TList,
+  listPromise: V1NamespacedList<TList>,
+): Promise<{
+  response: IncomingMessage;
+  body: TList;
+}> {
+  let continueToken: string | undefined = undefined;
+
+  const pretty = undefined;
+  const allowWatchBookmarks = undefined;
+  const fieldSelector = undefined;
+  const labelSelector = undefined;
+
+  let incomingMessage: IncomingMessage | undefined = undefined;
+
+  loop: while (true) {
+    try {
+      const listCall = await listPromise(
+        namespace,
+        pretty,
+        allowWatchBookmarks,
+        continueToken,
+        fieldSelector,
+        labelSelector,
+        PAGE_SIZE,
+      );
+      incomingMessage = listCall.response;
+
+      if (Array.isArray(listCall.body.items)) {
```

Comment:
is it ok if it is not an array?

Reply:
This also implicitly covers an `undefined` or `null`